src/retryOnEmpty.ts

```diff
@@ -140,7 +140,10 @@ async function retry(
   for (let index = 0; index < maxRetries; index++) {
     try {
       return await asyncFn();
-    } catch (err) {
+    } catch (err: any) {
+      if (err.code === -32000) {
+        throw err;
```

Comment:
Makes sense on the surface!

A few comments:

- `eth-json-rpc-middleware` uses our `@metamask/rpc-errors` package which provides an abstraction around possible RPC errors. It looks like there is a constant that represents this error code. I feel like ought to make use of it so that this isn't a "magic number".
- Since this is a `catch`, we don't really know that `err` has a code property — it's really `unknown` and not `any` — so to be safe, we should check for that property. This package also uses `@metamask/utils`, which has a `isValidJsonRpcError` utility function we can use.
- In your PR title, you say that you're handling a revert error. However, looking at `@metamask/rpc-errors`, [-32000 seems to be the code for an "invalid input" error](https://github.com/MetaMask/rpc-errors/blob/9200551fa80ab1f912704a2b346ef10f506878b0/src/error-constants.ts#L4). Therefore, if we truly want to handle a "revert" error, we might want to check the message as well as the code.

Putting this together, we might want to add a new function:

``` typescript
import { isValidJsonRpcError } from "@metamask/utils";
import type { JsonRpcError } from "@metamask/utils";
import { errorCodes } from "@metamask/rpc-errors";

function isExecutionRevertedError(error: unknown): error is JsonRpcError {
  return isValidJsonRpcError(error) &&
    error.code === errorCodes.rpc.invalidInput &&
    error.message === "execution reverted";
}
```

Then, here, we can say:

``` suggestion
    } catch (err: unknown) {
      if (isExecutionRevertedError(err)) {
        throw err;
```

Reply:
I added this improvement.